### PR TITLE
chore: create GitHub action to update the canary branch

### PR DESCRIPTION
**Pull request by targos**

### CODE REVIEW

**Comment by targos** [comment]
I created an empty `main` branch.
Here's a run on my fork: https://github.com/targos/node-v8/runs/1107906165?check_suite_focus=true

Things to note:
- We need to find a way to notify interested individuals when there's a failure, because I think errors in scheduled actions do not generate GitHub notifications
- Fetching V8 in node-core-utils takes a lot of time. Since I already clone the `lkgr` branch in the action, that step shouldn't be necessary and we can probably enhance ncu for that

**Comment by mmarchini** [comment]
> We need to find a way to notify interested individuals when there's a failure, because I think errors in scheduled actions do not generate GitHub notifications

It generates for the committer and for the author (maybe also for the pusher). Are you on the OpenJS Foundation slack? We could have a `nodejs-v8-canary` channel and then send the notifications there

**Comment by mmarchini** [comment]
> Fetching V8 in node-core-utils takes a lot of time. Since I already clone the lkgr branch in the action, that step shouldn't be necessary and we can probably enhance ncu for that

It doesn't respect the `--base-dir` option? Another option is to use a cache instead of checking out the repository (and let ncu check out like it does today, only the first run will take time to fetch V8).

**Comment by targos** [comment]
> It doesn't respect the --base-dir option? 

It does respect that option. The issue is that it fetches all branches with `git fetch origin`. That's not a problem with Jenkins because the workspaces are kept between runs so only the first run on a new machine would be slow.

**Comment by targos** [comment]
Using a Slack channel SGTM.

**Comment by mmarchini** [comment]
> It does respect that option. The issue is that it fetches all branches with git fetch origin. That's not a problem with Jenkins because the workspaces are kept between runs so only the first run on a new machine would be slow.

Ah, that sounds like a good use for cache then

**Comment by mmarchini** [comment]
For Slack there are many actions out there: https://github.com/marketplace?type=actions&query=slack. I have tested https://github.com/marketplace/actions/slack-notify but haven't tested it to "notify on failure". Other Actions look easier to use to notify on failure (or after every run, if that's what we want).

File: .github/workflows/update-canary.yml
Comment (gengjiawen) [inline .github/workflows/update-canary.yml:31]: Maybe install from github repo, we can have latest patch.

Comment (mmarchini) [reply .github/workflows/update-canary.yml:31]: I don't think that's necessary, if we need the newest version we can always make a ncu release (both myself and targos have access to do so)

**Comment by targos** [comment]
I created the `#nodejs-v8-canary` channel on Slack

**Comment by cclauss** [comment]
Sorry @targos I did not mean to undraft this!  Now I can not seem to redraft it.

File: .github/workflows/update-canary.yml
Comment (gengjiawen) [inline .github/workflows/update-canary.yml:5]: Also, other then schedule, I hope there is a way to manually trigger this like webook events.

Comment (mmarchini) [reply .github/workflows/update-canary.yml:5]: https://docs.github.com/en/actions/managing-workflow-runs/manually-running-a-workflow

I should really test this some time, it looks cool

Comment (mmarchini) [reply .github/workflows/update-canary.yml:5]: I wouldn't block this PR on it though, incremental improvements let us experiment faster :)

Comment (gengjiawen) [reply .github/workflows/update-canary.yml:5]: > I wouldn't block this PR on it though, incremental improvements let us experiment faster :)

Yeap. this shouldn't be a blocker. But we can keep jenkins job before it's done (remove schedule part in jenkins job).

**Review by crypt096 — APPROVED** [review]

**Comment by mmarchini** [comment]
@targos fyi I added SLACK_WEBHOOK Secret to the repo, so we can use it to notify the channel when sync fails

File: .github/workflows/update-canary.yml
Comment (targos) [reply .github/workflows/update-canary.yml:5]: I added the workflow_dispatch event

File: .github/workflows/update-canary.yml
Comment (targos) [inline .github/workflows/update-canary.yml:43]: Result:

image

File: .github/workflows/update-canary.yml
Comment (targos) [inline .github/workflows/update-canary.yml:39]:
```suggestion
```

File: README.md
Comment (targos) [inline README.md:1]:
```suggestion
# Node.js V8 canary
```

What do you think?

File: .github/workflows/update-canary.yml
Comment (targos) [inline .github/workflows/update-canary.yml:32]:
```suggestion
```

I'm not sure it's worth having this cache. It is only uploaded on the first successful run so with time we will lose the upside of it.
Currently, it reduces the run time by two minutes.

**Comment by targos** [comment]
Last try: https://github.com/targos/node-v8/runs/1800903471?check_suite_focus=true

Note that we apparently need a custom GitHub token with "workflows" permission, otherwise the push to `canary` fails if a workflow was updated in nodejs/node.
Appart from that, I think it's ready. The build took around 1hour. I don't know if we can optimize that somehow but the good news is that it's much faster than the Jenkins job!

**Comment by targos** [comment]
@mmarchini @gengjiawen would you like to have a new look?

**Review by gengjiawen — APPROVED** [review]
LGTM

**Comment by targos** [comment]
Starting with a manual run: https://github.com/nodejs/node-v8/actions/runs/737865140

**Pull request merged**
